### PR TITLE
Add API to get non-standard fields for contributor

### PR DESCRIPTION
**Pull request by TaiWilkin**
## Overview

Returns a list of all the field names that a contributor has submitted
either by API or list, minus our standard fields (country, name,
address, ppe_*, lat, lng).

Fields are parsed from list headers and from the keys of single-facility
API submissions.

Connects #1313 

## Demo

Screen Shot 2021-04-23 at 12 07 54 PM

## Testing Instructions

* Login as c2@example.com.
* Navigate to [facilities_create](http://localhost:8081/api/docs/#!/facilities/facilities_create), authorize as c2@example.com, and submit a facility with custom fields:

```
{
	"country": "China",
	"name": "Nantong Jackbeanie Headwear &amp; Garment Co. Ltd.",
	"address": "No.808,the third industry park,Guoyuan Town,Nantong 226500.",
	"ppe_product_types": ["Masks", "Gloves"],
	"ppe_contact_phone": "123-456-7890",
	"ppe_contact_email": "ppe@example.com",
	"ppe_website": "https://example.com/ppe",
	"test": "test data"
}
```
* Submit a second facility with custom fields, some repeating:
```
{
	"country": "China",
	"name": "Nantong Jackbeanie Headwear 2, Garment Co. Ltd.",
	"address": "No.809,the third industry park,Guoyuan Town,Nantong 226500.",
	"ppe_product_types": ["Masks", "Gloves"],
	"ppe_contact_phone": "123-456-7890",
	"ppe_contact_email": "ppe@example.com",
	"ppe_website": "https://example.com/ppe",
	"test": "test data!!",
	"custom_field": "field data"
}
```
* Submit and process a list with custom fields [test_list.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/6366722/OAR_Contributor_Template.csv)
* Navigate to [api/nonstandard-fields/](http://localhost:8081/api/nonstandard-fields/)
     - [x] You should see a list of custom fields with includes each of your custom fields exactly once

## Checklist

- [ ] `fixup!` commits have been squashed
- [ ] CI passes after rebase
- [ ] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

### CODE REVIEW

**Review by jwalgran — COMMENTED** [review]
This is a clear implementation of the feature as discussed and the "happy path" works as described. I did find some edge cases and a lingering bug that require us to make some fixes. I have mentioned them inline and here is a test case that exercises them which can be used for some test-driven fixing.

```python
class NonstandardFieldsApiTest(APITestCase):
    def setUp(self):
        self.url = reverse('nonstandard-fields-list')
        self.user_email = 'test@example.com'
        self.user_password = 'example123'
        self.user = User.objects.create(email=self.user_email)
        self.user.set_password(self.user_password)
        self.user.save()

        self.contributor = Contributor \
            .objects \
            .create(admin=self.user,
                    name='test contributor 1',
                    contrib_type=Contributor.OTHER_CONTRIB_TYPE)

        self.list = FacilityList \
            .objects \
            .create(header='country,name,address,extra_1',
                    file_name='one',
                    name='First List')

        self.list_source = Source \
            .objects \
            .create(facility_list=self.list,
                    source_type=Source.LIST,
                    is_active=True,
                    is_public=True,
                    contributor=self.contributor)

        self.list_item = FacilityListItem \
            .objects \
            .create(name='Item',
                    address='Address',
                    country_code='US',
                    row_index=1,
                    geocoded_point=Point(0, 0),
                    status=FacilityListItem.CONFIRMED_MATCH,
                    source=self.list_source)

        self.api_source = Source \
            .objects \
            .create(source_type=Source.SINGLE,
                    is_active=True,
                    is_public=True,
                    contributor=self.contributor)

        self.api_list_item = FacilityListItem \
            .objects \
            .create(name='Item',
                    address='Address',
                    country_code='US',
                    raw_data="{'country': 'US', 'name': 'Item', 'address': 'Address', 'extra_2': 'data'}",
                    row_index=1,
                    geocoded_point=Point(0, 0),
                    status=FacilityListItem.CONFIRMED_MATCH,
                    source=self.api_source)

    def test_nonstandard_fields(self):
        self.client.login(email=self.user_email,
                          password=self.user_password)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        content = json.loads(response.content)
        self.assertTrue(2, len(content))
        self.assertIn('extra_1', content)
        self.assertIn('extra_2', content)

    def test_doublequote_header(self):
        self.list.header='"country","name","address","extra_1"',
        self.list.save()

        self.client.login(email=self.user_email,
                          password=self.user_password)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        content = json.loads(response.content)
        self.assertIn('extra_1', content)

    def test_escaped_singlequote_in_api_data(self):
        self.api_list_item.raw_data = "{'country': 'US', 'name': 'Item', 'address': 'Address', 'extra_2': 'd\'ataé'}"
        self.api_list_item.save()

        self.client.login(email=self.user_email,
                          password=self.user_password)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        content = json.loads(response.content)
        self.assertIn('extra_2', content)

    def test_querydict_in_api_data(self):
        # Production data at the time this test was written includes a mix of
        # strigified dicts and stringified QueryDicts
        self.api_list_item.raw_data = "<QueryDict: {'name': ['Item'], 'country': ['US'], 'address': ['Address'], 'extra_2': ['data']}>"
        self.api_list_item.save()

        self.client.login(email=self.user_email,
                          password=self.user_password)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        content = json.loads(response.content)
        # We plan to only store raw data as CSV or JSON in the future. Legacy
        # data can be ignored
        self.assertNotIn('extra_2', content)
```

File: src/django/api/views.py [hunk 1]
```diff
@@ -3576,3 +3578,68 @@ def update(self, request, pk=None):
...
+
+def get_single_facility_fields(contributor):
+    single_facilities = FacilityListItem.objects.filter(
+                                            source__source_type='SINGLE',
+                                            source__contributor=contributor
+                                            ).values_list(
+                                            'raw_data', flat=True)
+    single_facility_fields = []
+    for single_facility in single_facilities:
+        fields = list(json.loads(single_facility.replace("'", '"')).keys())
```
[the remark this answers is not in the document]
Comment (TaiWilkin) [reply src/django/api/views.py]: I added a double try/except block to try handling the `raw_data` as JSON. If it fails, it jumps into the nested try/except to try changing the single quotes to double quotes and parsing that; failing that, it doesn't include the fields from that object in the response. This still doesn't handle all of our cases of legacy data, but it does allow at least some legacy data to be parsed, although it's not pretty. I'm open to alternative options for certain.

[the remark this answers is not in the document]
Comment (jwalgran) [reply src/django/api/views.py]: The nested fallback is a good approach. Thanks.

**Review by jwalgran — APPROVED** [review]
Thanks for working through these these problematic issues with real world data. I was no longer able to break it and confirmed the test instructions.

**Comment by TaiWilkin** [comment]
Thanks for the review and feedback!
